filecheck/src/FileCheck.hs

```diff
@@ -0,0 +1,205 @@
+        ]
+      , "pattern:"
+      , indent <> FCC.cmdContent c
+      , "match:"
+      , indentLines (FCC.matchText m)
+      ]
+
+instance Show FileCheckFailure where
+  show f =
+    let r = failureResult f in
+    Text.unpack $
+      Text.unlines
+      [ ""  -- a leading newline makes the output of Tasty more readable
+      , "Check failed! Passing checks:"
+      , printMatches (zip (failureCommands f) (toList (resultMatches r)))
+      , "Failing check:"
+      , Text.unwords
+        [ "❌"
+        , FCD.print (FCC.cmdDirective (failureFailedCommand f))
+        , "at"
+        , maybe "<unknown location>" FCP.printSpan (FCC.cmdSpan (failureFailedCommand f))
+        , "did not match text at"
+        , FCP.printLoc (resultLocation r)
+        ]
+      , "pattern:"
+      , indent <> FCC.cmdContent (failureFailedCommand f)
+      , "text:"
+      , indentLines (resultRemainder r)
+      ]
+
+instance X.Exception FileCheckFailure
+
+-- | The result of checking a sequence of t'Command's against some t'Output'
+data Result
+  = Result
+    { -- | t'Loc' after the last match
+      resultLocation :: Loc
+      -- | Successful 'Match'es
+    , resultMatches :: Seq Match
+      -- | Remaining text after the last match
+    , resultRemainder :: Text
+    }
+
+-- | Pretty-print a t'Result'
+printResult ::
+  [Command] ->
+  Result ->
+  Text
+printResult cmds r =
+  let matches = resultMatches r in
+  case drop (length matches) cmds of
+    [] -> printMatches (zip cmds (toList matches))
+    (failed : _) -> Text.pack (show (FileCheckFailure cmds failed r))
+
+-- | Match text against a sequence of t'Command's.
+check ::
+  [Command] ->
+  Output ->
+  Result
+check cmds (Output txt0) =
+  let loc0 = Loc (Just "<out>") (FCP.Pos 1 1) in
+  case foldl' go (Right (Result loc0 Seq.empty txt0)) cmds of
+    Left r -> r
+    Right r -> r
+  where
+  go (Right (Result loc ms txt)) cmd =
+    case FCC.match cmd loc txt of
+      Nothing -> Left (Result loc ms txt)
+      Just m ->
+        let loc' = FCP.endLoc (FCC.matchSpan m) in
+        Right (Result loc' (ms Seq.:|> m) (FCC.matchRemainder m))
+  go (Left r) _cmd = Left r
+
+-- | Helper, not exported
+checkResult ::
+  [Command] ->
+  Result ->
+  IO ([Command], Result)
+checkResult cmds r =
+  let matches = resultMatches r in
+  case drop (length matches) cmds of
+    [] -> pure (cmds, r)
+    (failed : _) -> X.throwIO (FileCheckFailure cmds failed r)
+
+-- | Like 'check', but throws t'FileCheckFailure' on failure.
+check' ::
+  HasCallStack =>
+  [Command] ->
+  Output ->
+  IO Result
+check' cmds out = snd <$> checkResult cmds (check cmds out)
+
+-- | Failure to parse a t'Command'
+data CommandParseFailure = CommandParseFailure Text
+  deriving Show
+
+instance X.Exception CommandParseFailure
+
+-- | Parse t'Command's from comments embedded in a file, and use them to check
+-- an t'Output'.
+--
+-- Throws t'CommandParseFailure' if the commands cannot be parsed.
```

Comment:
I tried looking to see where `CommandParseFailure` is thrown, but I couldn't find it. Am I missing something?

Reply:
Indeed, this is outdated. Thanks for the catch!